x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java

```diff
@@ -43,6 +44,26 @@ protected AbstractBlock(int positionCount, @Nullable int[] firstValueIndexes, @N
         this.firstValueIndexes = firstValueIndexes;
         this.nullsMask = nullsMask == null || nullsMask.isEmpty() ? null : nullsMask;
         assert nullsMask != null || firstValueIndexes != null : "Create VectorBlock instead";
+        assert assertInvariants();
+    }
+
+    private boolean assertInvariants() {
+        if (firstValueIndexes != null) {
+            assert firstValueIndexes.length == getPositionCount() + 1;
+            for (int i = 0; i < getPositionCount(); i++) {
+                assert (firstValueIndexes[i + 1] - firstValueIndexes[i]) >= 0;
+            }
+        }
+        if (nullsMask != null) {
+            assert nullsMask.nextSetBit(getPositionCount() + 1) == -1;
```

Comment:
This is new. This was assumed in `shiftNullsToExpandedPositions` may end up accessing `firstValueIndexes` out of bounds.